nexus/src/app/instance_network.rs

```diff
@@ -344,47 +343,71 @@ impl super::Nexus {
             .instance_lookup_external_ips(&opctx, instance_id)
             .await?;
 
-        if let Some(wanted_index) = ip_index_filter {
-            if let None = ips.get(wanted_index) {
+        let (ips_of_interest, must_all_be_attached) = if let Some(wanted_id) =
+            ip_filter
+        {
+            if let Some(ip) = ips.iter().find(|v| v.id == wanted_id) {
+                (std::slice::from_ref(ip), false)
+            } else {
                 return Err(Error::internal_error(&format!(
-                    "failed to find external ip address at index: {}",
-                    wanted_index
+                    "failed to find external ip address with id: {wanted_id}, saw {ips:?}",
                 )));
             }
+        } else {
+            (&ips[..], true)
+        };
+
+        // This is performed so that an IP attach/detach will block the
+        // instance_start saga. Return service unavailable to indicate
+        // the request is retryable.
+        if must_all_be_attached
```

Comment:
Re-ensuring an existing NAT entry for the RPW is a no-op, and should only cost a row lookup. So there's nothing wrong with always pushing all `Attached` IPs apart from wasted effort.

We do still require an override for at least one other entry in an arbitrary state, because of saga unwind of attach/detach: an IP might be `Detaching`, unwind when informing OPTE, and then need to be specifically readded to NAT. Similarly an `Attaching` IP might unwind at the same step, and need to be removed from NAT. As a result, we don't want other executing sagas to try to do anything with IPs in a state they don't expect or IPs they don't 'own'. We'd also need to ensure that instance_start sees "all attached", and that attach/detach don't care about other IPs' states because they can safely run concurrently for different IP objects on the same instance.

Unrelated: Looking over this again I've also seen that `ensure_nat_entry` is needlessly duplicated per IP, which is again a no-op. This is also indirectly duplicated today in `main` (instance_start: `for switch in boundary_switches {... instance_ensure_dpd_config ...}` -> `for ip in ips { ... ensure_nat_entry ... }`), but we can fix this here now.